File: src/lib/llm/prompt-builder.ts
import fs from 'fs'
import path from 'path'
import { ChatMode } from '@/types/chat'

const PROMPTS_DIR = path.join(process.cwd(), 'src', 'lib', 'prompts')

// Read prompt templates at module load time
const MOCK_DATA_PROMPT = fs.readFileSync(
  path.join(PROMPTS_DIR, 'mock-data.md'),
  'utf-8'
)

const WIDGET_CODE_PROMPT = fs.readFileSync(
  path.join(PROMPTS_DIR, 'widget-code.md'),
  'utf-8'
)

const WIDGET_REFINE_SYSTEM_PROMPT = fs.readFileSync(
  path.join(PROMPTS_DIR, 'widget-refine-system.md'),
  'utf-8'
)

const A2UI_WIDGET_PROMPT = fs.readFileSync(
  path.join(PROMPTS_DIR, 'a2ui-widget.md'),
  'utf-8'
)

const A2UI_REFINE_SYSTEM_PROMPT = fs.readFileSync(
  path.join(PROMPTS_DIR, 'a2ui-refine-system.md'),
  'utf-8'
)

export function buildMockDataPrompt(
  mode: ChatMode,
  input: string,
  appMetadata?: Record<string, any>
): string {
  let prompt = MOCK_DATA_PROMPT

  // Replace MODE placeholder
  prompt = prompt.replace('{{MODE}}', mode)

  // Replace INPUT_CONTENT placeholder
  if (mode === 'appId' && appMetadata) {
    const content = `App Metadata:\n${JSON.stringify(appMetadata, null, 2)}`
    prompt = prompt.replace('{{INPUT_CONTENT}}', content)
  } else {
    const content = `App Description:\n${input}`
    prompt = prompt.replace('{{INPUT_CONTENT}}', content)
  }

  return prompt
}

export function buildWidgetPrompt(
  mockData: Record<string, any>,
  uiStylePrompt: string
): string {
  let prompt = WIDGET_CODE_PROMPT

  // Replace MOCK_DATA_SCHEMA placeholder
  prompt = prompt.replace(
    '{{MOCK_DATA_SCHEMA}}',
    JSON.stringify(mockData, null, 2)
  )

  // Replace UI_STYLE_PROMPT placeholder
  prompt = prompt.replace('{{UI_STYLE_PROMPT}}', uiStylePrompt)

  return prompt
}

export function buildWidgetRefineMessages(
  previousCode: string,
  refinePrompt: string
): { system: string; user: string } {
  return {
    system: WIDGET_REFINE_SYSTEM_PROMPT,
    user: `${previousCode}\n\n${refinePrompt}`,
  }
}

export function buildA2UIWidgetPrompt(
  mockData: Record<string, any>,
  uiStylePrompt: string
): string {
  let prompt = A2UI_WIDGET_PROMPT

  prompt = prompt.replace(
    '{{MOCK_DATA_SCHEMA}}',
    JSON.stringify(mockData, null, 2)
  )

  prompt = prompt.replace('{{UI_STYLE_PROMPT}}', uiStylePrompt)

  return prompt
}

export function buildA2UIRefineMessages(
  previousSchema: string,
  refinePrompt: string
): { system: string; user: string } {
  return {
    system: A2UI_REFINE_SYSTEM_PROMPT,
    user: `${previousSchema}\n\n${refinePrompt}`,
  }
}

File: training/src/prompt_builder.py
"""Python port of src/lib/llm/prompt-builder.ts — builds prompts for widget and mock data generation."""

import json
from pathlib import Path

PROMPTS_DIR = Path(__file__).parent.parent / "prompts"

# Read templates once at module load
MOCK_DATA_PROMPT = (PROMPTS_DIR / "mock-data.md").read_text()
WIDGET_CODE_PROMPT = (PROMPTS_DIR / "widget-code.md").read_text()


def build_mock_data_prompt(
    mode: str,
    input_content: str,
    app_metadata: dict | None = None,
) -> str:
    """Build the mock data generation prompt.

    Mirrors buildMockDataPrompt() from prompt-builder.ts.
    """
    prompt = MOCK_DATA_PROMPT
    prompt = prompt.replace("{{MODE}}", mode)

    if mode == "appId" and app_metadata:
        content = f"App Metadata:\n{json.dumps(app_metadata, indent=2)}"
    else:
        content = f"App Description:\n{input_content}"

    prompt = prompt.replace("{{INPUT_CONTENT}}", content)
    return prompt


def build_widget_prompt(mock_data: dict, ui_style_prompt: str) -> str:
    """Build the widget code generation prompt.

    Mirrors buildWidgetPrompt() from prompt-builder.ts.
    """
    prompt = WIDGET_CODE_PROMPT
    prompt = prompt.replace("{{MOCK_DATA_SCHEMA}}", json.dumps(mock_data, indent=2))
    prompt = prompt.replace("{{UI_STYLE_PROMPT}}", ui_style_prompt)
    return prompt
